### ISSUE TITLE
NSExceptionRequiresForwardSecrecy Issue

### ISSUE THREAD

**Issue by craigj2** [forcedotcom/SalesforceMobileSDK-iOS#1317]
Hello,

I am creating a new sample app using forceios create.

When I build using cordova build, the NSExceptionRequiresForwardSecrecy line is not copied to my *-info.plist file. This causes an SSL error when the app tries to display the salesforce login screen.

My plist file looks like this:

```
    <key>NSAppTransportSecurity</key>
    <dict>
      <key>NSExceptionDomains</key>
      <dict>
        <key>salesforce.com</key>
        <dict>
          <key>NSIncludesSubdomains</key>
          <true/>
        </dict>
        <key>force.com</key>
        <dict>
          <key>NSIncludesSubdomains</key>
          <true/>
        </dict>
        <key>localhost</key>
        <dict/>
      </dict>
      <key>NSAllowsArbitraryLoads</key>
      <true/>
    </dict>
```

Rather than as specified in com.salesforce/plugin.xml which looks like this:

```
    <config-file target="*-Info.plist" parent="NSAppTransportSecurity">
      <dict>
        <key>NSExceptionDomains</key>
        <dict>
          <key>salesforce.com</key>
          <dict>
            <key>NSIncludesSubdomains</key>
            <true/>
            <key>NSExceptionRequiresForwardSecrecy</key>
            <false/>
          </dict>
          <key>force.com</key>
          <dict>
            <key>NSIncludesSubdomains</key>
            <true/>
            <key>NSExceptionRequiresForwardSecrecy</key>
            <false/>
          </dict>
        </dict>
      </dict>
    </config-file>
```

This is a newly created project, any ideas what might have gone wrong?

Thanks!

**Comment by toddhalfpenny** [comment]
Hi @craigj2, I think this may be a bug (which has been fixed) in cordova. What version of cordova do you have?

**Comment by craigj2** [comment]
Hi @toddhalfpenny, 
Thanks for your reply! I'm using 5.4.1 on OSX.

**Comment by toddhalfpenny** [comment]
Ah actually which version of cordova-ios have you got?
I think this was the patch that fixed it IIRC https://github.com/apache/cordova-ios/commit/8d4196f77cd0eaff8fa802a4d5173d867da90940

**Comment by craigj2** [comment]
Thanks for your help @toddhalfpenny 

I ran cordova platforms list and it returned

'''
Installed platforms: ios 3.9.2
'''

I have ran cordova platform update ios, And I have also ran npm update -g cordova is there anything else I'm missing? 

Thanks Again!

**Comment by toddhalfpenny** [comment]
OK, I think you need cordova-ios 4.0.0 or 4.0.1 (sorry not at my mac at the mo but I think those builds had the fix in). Once you have this can you check again?

**Comment by craigj2** [comment]
Thanks again @toddhalfpenny I tried adding the specific version using

```
cordova platform update ios@4.0.1
```

But now the build fails with

```
The following build commands failed:
	CompileC build/xxx.build/Debug-iphonesimulator/xxx.build/Objects-normal/i386/MainViewController.o xxx/Classes/MainViewController.m normal i386 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
ERROR building one of the platforms: Error code 65 for command: xcodebuild with args: -xcconfig,/Users/xxx/xxx/xxx/platforms/ios/cordova/build-debug.xcconfig,-project,xxx.xcodeproj,ARCHS=i386,-target,xxx,-configuration,Debug,-sdk,iphonesimulator,build,VALID_ARCHS=i386,CONFIGURATION_BUILD_DIR=/Users/xxx/xxx/xxx/platforms/ios/build/emulator,SHARED_PRECOMPS_DIR=/Users/xxx/xxx/xxx/platforms/ios/build/sharedpch
You may not have the required environment or OS to build this project
```

**Comment by craigj2** [comment]
This same issue is happening o  my colleagues brand new mac with a fresh install of everything. Whilst it may be a cordova issue, it looks like in its current form, salesforce hybrid apps for iOS don't work.

The only work around I have found is to manually edit the *-Info.plist file, which is overwritten each time I build.

Can anyone confirm this? Or suggest any other work around? Or does anybody have any tips for using cordova-ios 4.0.1, as I just cant get it to work and keep receiving the error above.

Thanks for your help.

**Comment by khawkins** [comment]
`cordova platform update` does not appear to work to update from 3.9 to 4.x, for any Cordova app.  But in this case, it wouldn't matter, since the Mobile SDK does not yet support Cordova iOS 4.0.

These settings updates in the Info.plist file seem to be driven by the whitelist entries in the plugin configuration, not the explicit app transport security configuration.  If I consume a plugin that has the whitelist entries commented out, the proper app transport settings get implemented from the plugin configuration.

Right now, the only way to work around the situation is the way you suggest: updating your Info.plist file whenever necessary.  We'll have to see what the best approach is to fix this in the plugin itself, prior to the Cordova iOS 4.0 support implementation.

**Comment by khawkins** [comment]
Note that we'll be disabling our whitelist entries in the next patch release of the plugin.  See [this pull request](https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin/pull/142) for more information.

**Comment by bhariharan** [comment]
We have fixed this issue in our ```4.0.2``` patch release, which is live now.
